pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -144,10 +207,79 @@ func GroupHibernationSchedulesByLocation(schedules []gardencorev1beta1.Hibernati
 			locationID = time.UTC.String()
 		}
 
-		locationToSchedules[locationID] = append(locationToSchedules[locationID], schedule)
+		location, err := time.LoadLocation(locationID)
+		if err != nil {
+			return nil, err
+		}
+		if schedule.Start != nil {
+			parsed, err := cron.ParseStandard(*schedule.Start)
+			if err != nil {
+				return nil, err
+			}
+			parsedHibernationSchedules = append(parsedHibernationSchedules,
+				ParsedHibernationSchedule{Location: location, Schedule: parsed, Enabled: true},
+			)
+		}
+		if schedule.End != nil {
+			parsed, err := cron.ParseStandard(*schedule.End)
+			if err != nil {
+				return nil, err
+			}
+			parsedHibernationSchedules = append(parsedHibernationSchedules,
+				ParsedHibernationSchedule{Location: location, Schedule: parsed, Enabled: false},
+			)
+		}
+	}
+
+	return parsedHibernationSchedules, nil
+}
+
+// nextHibernationTimeDuration returns the time duration after which to requeue the shoot based on the hibernation schedules and current time.
+func nextHibernationTimeDuration(schedules []ParsedHibernationSchedule, now time.Time) time.Duration {
+	var timestamps []time.Time
+	for _, schedule := range schedules {
+		ts := schedule.Next(now)
+		timestamps = append(timestamps, ts)
+	}
+
+	sort.Slice(timestamps, func(i, j int) bool {
+		return timestamps[i].Before(timestamps[j])
+	})
+
+	duration := timestamps[0].Sub(now)
+	return duration
+}
+
+// getScheduleWithMostRecentTime returns the ParsedHibernationSchedule that contains the schedule with the most recent previous execution time.
+func getScheduleWithMostRecentTime(schedules []ParsedHibernationSchedule, now time.Time, shoot *gardencorev1beta1.Shoot) *ParsedHibernationSchedule {
+	var startTime time.Time
+	if shoot.Status.LastHibernationTriggerTime != nil {
+		startTime = shoot.Status.LastHibernationTriggerTime.Time
+	} else {
+		// if the shoot has just been created or has never been hibernated, use the creation timestamp
+		startTime = shoot.CreationTimestamp.Time
+	}
+
+	var scheduleWithMostRecentTime *ParsedHibernationSchedule
+	for i := 0; i < len(schedules); i++ {
+		cur := schedules[i].Prev(startTime, now)
+		if cur == nil {
+			continue
+		}
+		if scheduleWithMostRecentTime == nil {
+			scheduleWithMostRecentTime = &schedules[i]
+			continue
+		}
+		mostRecentTime := scheduleWithMostRecentTime.Prev(startTime, now)
+		if mostRecentTime == nil {
+			continue
+		}
+		if cur.After(*mostRecentTime) {
+			scheduleWithMostRecentTime = &schedules[i]
+		}
```

Comment:
```suggestion
	for _, schedule := range schedules {
		cur := schedule.Prev(startTime, now)
		if cur == nil {
			continue
		}

		if scheduleWithMostRecentTime == nil {
			scheduleWithMostRecentTime = &schedule
			continue
		}

		mostRecentTime := scheduleWithMostRecentTime.Prev(startTime, now)
		if mostRecentTime == nil {
			continue
		}

		if cur.After(*mostRecentTime) {
			scheduleWithMostRecentTime = &schedule
		}
```

Reply:
Not sure if this would work. `scheduleWithMostRecentTime` gets the address of the `schedule` variable and its value changes on each iteration. In the end `scheduleWithMostRecentTime` will be a pointer to a value that is equal to the last element of the `schedules` array.

Reply:
fair enough, but you could still use the `for x,y := range z {}` notation instead of the C++ like `for i := 0; i < len(schedules); i++ {` :)